course/src/isometry.cc

```diff
@@ -0,0 +1,245 @@
+        double m02 = sy;
+        double m10 = sx * sy * cz + cz * cx;
+        double m11 = -sx * sy * sz + cx * cz;
+        double m12 = -sx * sy;
+        double m20 = -cx * sy * cz + sx * sz;
+        double m21 = cx * sy * sz + sx * cz;
+        double m22 = cx * cy;
+
+        Vector3 vv1 = Vector3(m00, m01, m02);
+        Vector3 vv2 = Vector3(m10, m11, m12);
+        Vector3 vv3 = Vector3(m20, m21, m22);
+
+        return Isometry(Vector3::kZero, Matrix3(vv1, vv2, vv3));
+    }
+
+    Vector3 Isometry::transform(const Vector3 &v) const
+    {
+        return ((*this) * v);
+    }
+
+    Vector3 Isometry::transform(const std::initializer_list<double> &list) const
+    {
+        if (list.size() == 3)
+        {
+            try
+            {
+                return ((*this) * Vector3(list));
+            }
+            catch (const std::exception &ex)
+            {
+                std::cerr << "Error occurred: " << ex.what() << std::endl;
+            }
+        }
+
+        return Vector3();
+    }
+
+    Isometry Isometry::inverse(void) const
+    {
+        return Isometry(vector3_isometry * (-1), matrix3_isometry);
+    }
+
+    Isometry Isometry::compose(const Isometry &I1) const
+    {
+        return Isometry((*this) * I1);
+    }
+
+    bool operator==(const Isometry &p, const Isometry &q)
+    {
+        for (size_t i = 0; i < 2; ++i)
+        {
+            if (p.vector3_isometry[i] != p.vector3_isometry[i])
+                return false;
+            if (p.matrix3_isometry[i] != p.matrix3_isometry[i])
+                return false;
+        }
+        return true;
+    }
+
+    bool operator!=(const Isometry &p, const Isometry &q)
+    {
+        return !(p == q);
+    }
+
+    Vector3 operator*(const Isometry &i, const Vector3 &v)
+    {
+        double rx = i.matrix3_isometry.row(0).dot(v) + i.vector3_isometry.x();
+        double ry = i.matrix3_isometry.row(1).dot(v) + i.vector3_isometry.y();
+        double rz = i.matrix3_isometry.row(2).dot(v) + i.vector3_isometry.z();
+        return Vector3(rx, ry, rz);
+    }
+
+    Isometry operator*(const Isometry &i, const Isometry &i2)
+    {
+        Isometry m;
+        Matrix3 p = i.matrix3_isometry;
+        Matrix3 q = i2.matrix3_isometry;
+
+        Vector3 a, b;
+
+        for (size_t i = 0; i < 3; ++i)
+        {
+            a.x() = p.row(i).x();
+            a.y() = p.row(i).y();
+            a.z() = p.row(i).z();
+            for (size_t j = 0; j < 3; ++j)
+            {
+                b = q.col(j);
+                m.matrix3_isometry[i][j] = a.dot(b);
+            }
+        }
+        m.vector3_isometry = i.vector3_isometry + i2.vector3_isometry;
+        return m;
+    }
+
+    std::ostream &operator<<(std::ostream &os, const Isometry &p)
+    {
+        os.precision(9);
+        os << "[T: " << p.vector3_isometry << ", ";
+        os << "R:" << p.matrix3_isometry << "]";
+        return os;
+    }
+
+    Isometry &Isometry::operator+=(const Isometry &q)
```

Comment:
Adding or subtracting isometries together is not really that makes sense, since the result is most likely not a valid transformation.

Reply:
Done